tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc

```diff
@@ -0,0 +1,2918 @@
+      tfl_relu_op.getResult().getType().dyn_cast<RankedTensorType>();
+  // Not a ranked tensor output
+  if (!input_type || !output_type) return failure();
+
+  bool input_is_qtype =
+      input_type.getElementType().isa<mlir::quant::UniformQuantizedType>();
+  bool output_is_qtype =
+      output_type.getElementType().isa<mlir::quant::UniformQuantizedType>();
+
+  if (input_is_qtype != output_is_qtype) {
+    op->emitOpError(
+        "ConvertTFLReluOp: input/output tensor should "
+        "be all quantized or all native");
+    return failure();
+  }
+
+  Value output;
+  if (output_is_qtype) {
+    auto rescale_type =
+        RankedTensorType::get(output_type.getShape(), rewriter.getI32Type());
+    auto input_qtype = input_type.getElementType()
+                           .dyn_cast<mlir::quant::UniformQuantizedType>();
+    auto output_qtype = output_type.getElementType()
+                            .dyn_cast<mlir::quant::UniformQuantizedType>();
+
+    auto op1_rescale_in = buildRescaleToInt32(rewriter, op, tfl_relu_op.x(),
+                                              1.0f, input_qtype.getZeroPoint());
+    auto op2_relun_op1 = rewriter.create<tosa::ReluNOp>(
+        op->getLoc(), rescale_type, op1_rescale_in,
+        rewriter.getI64IntegerAttr(std::numeric_limits<int32_t>::max()),
+        rewriter.getF32FloatAttr(0.0f));
+    auto op3_rescale_op2 = buildRescaleFromInt32(
+        rewriter, op, output_type, op2_relun_op1.getResult(), 1.0f,
+        output_qtype.getZeroPoint());
+
+    output = op3_rescale_op2;
+  } else {
+    auto op1_relun_in = rewriter.create<tosa::ReluNOp>(
+        op->getLoc(), output_type, tfl_relu_op.x(),
+        rewriter.getI64IntegerAttr(0),
+        rewriter.getF32FloatAttr(std::numeric_limits<float>::max()));
+
+    output = op1_relun_in.getResult();
+  }
+
+  rewriter.replaceOp(op, {output});
+  return success();
+}
+
+LogicalResult ConvertTFLRelu6Op::matchAndRewrite(
+    Operation* op, PatternRewriter& rewriter) const {
+  auto tfl_relu6_op = cast<TFL::Relu6Op>(op);
+
+  auto input_type = tfl_relu6_op.x().getType().dyn_cast<RankedTensorType>();
+  auto output_type =
+      tfl_relu6_op.getResult().getType().dyn_cast<RankedTensorType>();
+  // Not a ranked tensor output
+  if (!input_type || !output_type) return failure();
+
+  bool input_is_qtype =
+      input_type.getElementType().isa<mlir::quant::UniformQuantizedType>();
+  bool output_is_qtype =
+      output_type.getElementType().isa<mlir::quant::UniformQuantizedType>();
+
+  if (input_is_qtype != output_is_qtype) {
+    op->emitOpError(
+        "ConvertTFLRelu6Op: input/output tensor should "
+        "be all quantized or all native");
+    return failure();
+  }
+
+  Value output;
+  if (output_is_qtype) {
+    auto rescale_type =
+        RankedTensorType::get(output_type.getShape(), rewriter.getI32Type());
+    auto input_qtype = input_type.getElementType()
+                           .dyn_cast<mlir::quant::UniformQuantizedType>();
```

Comment:
Hoist out of condition and then if() on them.

Reply:
Given that we check whether input_is_qtype matches output_is_qtype above this code, and the two *_is_qtype perform the check, it seems the simpler option is to change the if condition to if(output_is_qtype && input_is_qtype).